Text-to-SQL Project.

Study of one of the IBM Granite model, here the 20B parameters code base 8K, for the generation of SQL queries from question in natural language.

Partie 1.

Load the model and the dataset

In [ ]:
!pip install transformers bitsandbytes accelerate datasets outlines scikit-learn

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 8.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 87.6/87.6 kB 7.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 343.6/343.6 kB 24.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 11.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 15.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/134.8 kB 12.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 912.7/912.7 kB 36.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 3.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 111.0/111.0 kB 10.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 101.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 194.1/194.1 kB 17.2 MB/s eta 0:00:00
  Attempting uninstall: fsspec
    Found existing installation: fsspec 2024.10.0
    Uninstalling fsspec-2024.10.0:
   

In [ ]:
from transformers import (AutoModelForCausalLM,BitsAndBytesConfig, AutoTokenizer)

import torch
import os
from dotenv import load_dotenv

load_dotenv()


hf_token = os.getenv("HFTOKEN")
model_path = "ibm-granite/granite-20b-code-base-8k"
#Load the tokenizer associated on the model
tokenizer = AutoTokenizer.from_pretrained(model_path, token=hf_token)
# Prevent some transformers specific issues.
tokenizer.use_default_system_prompt = False
tokenizer.pad_token_id = tokenizer.eos_token_id

# We want to use 4bit quantization to save memory
quantization_config = BitsAndBytesConfig(
    load_in_8bit=False, load_in_4bit=True
)

#Load the model
model = AutoModelForCausalLM.from_pretrained(model_path,quantization_config=quantization_config,
                                             device_map={"": 0},
                                             torch_dtype=torch.bfloat16,
                                             token=hf_token)
#Set the model for evaluation
model.eval()





tokenizer_config.json:   0%|          | 0.00/4.11k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.06M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/1.02k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/644 [00:00<?, ?B/s]

model.safetensors.index.json:   0%|          | 0.00/47.0k [00:00<?, ?B/s]

model-00001-of-00009.safetensors:   0%|          | 0.00/4.95G [00:00<?, ?B/s]

model-00002-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00003-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00004-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00005-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00006-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00007-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00008-of-00009.safetensors:   0%|          | 0.00/4.93G [00:00<?, ?B/s]

model-00009-of-00009.safetensors:   0%|          | 0.00/680M [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/9 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/132 [00:00<?, ?B/s]

GPTBigCodeForCausalLM(
  (transformer): GPTBigCodeModel(
    (wte): Embedding(49152, 6144)
    (wpe): Embedding(8192, 6144)
    (drop): Dropout(p=0.1, inplace=False)
    (h): ModuleList(
      (0-51): 52 x GPTBigCodeBlock(
        (ln_1): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (attn): GPTBigCodeSdpaAttention(
          (c_attn): Linear4bit(in_features=6144, out_features=6400, bias=True)
          (c_proj): Linear4bit(in_features=6144, out_features=6144, bias=True)
          (attn_dropout): Dropout(p=0.1, inplace=False)
          (resid_dropout): Dropout(p=0.1, inplace=False)
        )
        (ln_2): LayerNorm((6144,), eps=1e-05, elementwise_affine=True)
        (mlp): GPTBigCodeMLP(
          (c_fc): Linear4bit(in_features=6144, out_features=24576, bias=True)
          (c_proj): Linear4bit(in_features=24576, out_features=6144, bias=True)
          (act): GELUActivation()
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
    )
    (ln_f): Laye

In [ ]:
from transformers import GenerationConfig

generation_config = GenerationConfig(
  max_new_tokens = 196,
  min_new_tokens = 10,
  do_sample=False,
  temperature=0.3,
  repetition_penalty=1.0,
  eos_token_id=tokenizer.eos_token_id,
  pad_token_id=tokenizer.pad_token_id,
)

/usr/local/lib/python3.10/dist-packages/transformers/generation/configuration_utils.py:628: UserWarning: `do_sample` is set to `False`. However, `temperature` is set to `0.3` -- this flag is only used in sample-based generation modes. You should set `do_sample=True` or unset `temperature`. This was detected when initializing the generation config instance, which means the corresponding file may hold incorrect parameterization and should be fixed.
  warnings.warn(


In [ ]:
from datasets import load_dataset

#Dataset: https://huggingface.co/datasets/xlangai/spider?row=17
ds = load_dataset("xlangai/spider")

print(ds)



/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


README.md:   0%|          | 0.00/5.51k [00:00<?, ?B/s]

train-00000-of-00001.parquet:   0%|          | 0.00/831k [00:00<?, ?B/s]

validation-00000-of-00001.parquet:   0%|          | 0.00/126k [00:00<?, ?B/s]

Generating train split:   0%|          | 0/7000 [00:00<?, ? examples/s]

Generating validation split:   0%|          | 0/1034 [00:00<?, ? examples/s]

DatasetDict({
    train: Dataset({
        features: ['db_id', 'query', 'question', 'query_toks', 'query_toks_no_value', 'question_toks'],
        num_rows: 7000
    })
    validation: Dataset({
        features: ['db_id', 'query', 'question', 'query_toks', 'query_toks_no_value', 'question_toks'],
        num_rows: 1034
    })
})


In [ ]:
#Only looking for the validation test of the dataset
examples = [
    {"question": ex["question"], "SQL_query": ex["query"], "db_id": ex["db_id"]}
    for ex in ds['validation'].to_list()
]

#print(examples[0])

# Split examples and keep only a few samples to have short computation time.
test, train = examples[:100], examples[100:500]
print(f"Train len {len(train)}. Test len {len(test)}")
print(f"First example of test:\n{test[0]}")

Train len 400. Test len 100
First example of test:
{'question': 'How many singers do we have?', 'SQL_query': 'SELECT count(*) FROM singer', 'db_id': 'concert_singer'}


Partie 2 : Test the model with Zero shot prompt

In [ ]:
prompt = """
You are an expert in SQL and database querying.
 Your task is to convert natural language questions into optimized SQL queries.
Always return only the SQL query, without any explanation or additional text.
Here is the question: [{sentence}].
""".strip()

In [ ]:
def generate(prompt, llm=model, generation_config=generation_config):

  # Create turns with the given prompt

  # Tokenize turns, tokenize the text
  input_tokens = tokenizer(prompt, return_tensors='pt').to('cuda')
  input_ids=input_tokens['input_ids']

  # Ensure we don't use gradient to save memory space and computation time.
  #Generate output tokens
  with torch.no_grad():
    outputs = llm.generate(input_ids).to('cuda')

  #decode output token into text

  # Recover and decode answer.
  answer_tokens = outputs[0, input_ids.shape[1]:-1]
  #decode output token into text
  #.strip() delete space useless
  return tokenizer.decode(answer_tokens).strip()

In [ ]:
# Test your code

example = test[0]

#prompt = zeroshot_prompt.format(sentence=example["text"])
prompt = prompt.format(sentence=example["question"])

answer = generate(prompt)

print("##### Example 0 #####")
print(f"# {prompt}")
print(f"# {answer}")
print(f"  SQL query: {example['SQL_query']}")

##### Example 0 #####
# You are an expert in SQL and database querying.
 Your task is to convert natural language questions into optimized SQL queries.
Always return only the SQL query, without any explanation or additional text.
Here is the question: [How many singers do we have?].
# ASSISTANT: SELECT COUNT(*) FROM Singers;
  SQL query: SELECT count(*) FROM singer


Granite Code 20b has a context window size of 8192, which means that we can use up to 8192 tokens to set the context. In our case, we can pass information about the database schema in order to get more accurate answers from the model.